File: src/ts/impl/polygon_util.ts
import * as log from 'loglevel';
import * as PolyK from 'polyk';
import Vector from '../vector';
import * as jsts from 'jsts';

export default class PolygonUtil {
    private static geometryFactory = new jsts.geom.GeometryFactory(); // JSTS 几何工厂，用于创建几何对象

    /**
     * 切割矩形，返回最小的多边形
     * @param origin 矩形的起点
     * @param worldDimensions 矩形的宽高
     * @param p1 切割线的起点
     * @param p2 切割线的终点
     */
    public static sliceRectangle(origin: Vector, worldDimensions: Vector, p1: Vector, p2: Vector): Vector[] {
        const rectangle = [
            origin.x, origin.y,
            origin.x + worldDimensions.x, origin.y,
            origin.x + worldDimensions.x, origin.y + worldDimensions.y,
            origin.x, origin.y + worldDimensions.y,
        ];
        const sliced = PolyK.Slice(rectangle, p1.x, p1.y, p2.x, p2.y).map(p => PolygonUtil.polygonArrayToPolygon(p));
        const minArea = PolygonUtil.calcPolygonArea(sliced[0]);

        if (sliced.length > 1 && PolygonUtil.calcPolygonArea(sliced[1]) < minArea) {
            return sliced[1];
        }

        return sliced[0];
    }

    /**
     * 用于创建海洋多边形
     * @param origin 矩形的起点
     * @param worldDimensions 矩形的宽高
     * @param line 切割线
     */
    public static lineRectanglePolygonIntersection(origin: Vector, worldDimensions: Vector, line: Vector[]): Vector[] {
        const jstsLine = PolygonUtil.lineToJts(line); // 将线转换为 JSTS 几何对象
        const bounds = [
            origin,
            new Vector(origin.x + worldDimensions.x, origin.y),
            new Vector(origin.x + worldDimensions.x, origin.y + worldDimensions.y),
            new Vector(origin.x, origin.y + worldDimensions.y),
        ];
        const boundingPoly = PolygonUtil.polygonToJts(bounds); // 创建矩形的 JSTS 多边形
        const union = boundingPoly.getExteriorRing().union(jstsLine); // 合并边界和线
        const polygonizer = new (jsts.operation as any).polygonize.Polygonizer(); // 创建多边形化器
        polygonizer.add(union);
        const polygons = polygonizer.getPolygons();

        let smallestArea = Infinity;
        let smallestPoly;
        for (let i = polygons.iterator(); i.hasNext();) {
            const polygon = i.next();
            const area = polygon.getArea();
            if (area < smallestArea) {
                smallestArea = area;
                smallestPoly = polygon;
            }
        }

        if (!smallestPoly) return [];
        return smallestPoly.getCoordinates().map((c: any) => new Vector(c.x, c.y));
    }

    /**
     * 计算多边形的面积
     * @param polygon 多边形顶点数组
     */
    public static calcPolygonArea(polygon: Vector[]): number {
        let total = 0;

        for (let i = 0; i < polygon.length; i++) {
            const addX = polygon[i].x;
            const addY = polygon[i == polygon.length - 1 ? 0 : i + 1].y;
            const subX = polygon[i == polygon.length - 1 ? 0 : i + 1].x;
            const subY = polygon[i].y;

            total += (addX * addY * 0.5);
            total -= (subX * subY * 0.5);
        }

        return Math.abs(total);
    }

    /**
     * 递归地将多边形按最长边分割，直到满足最小面积条件
     * @param p 多边形顶点数组
     * @param minArea 最小面积
     */
    public static subdividePolygon(p: Vector[], minArea: number): Vector[][] {
        const area = PolygonUtil.calcPolygonArea(p);
        if (area < 0.5 * minArea) {
            return [];
        }
        const divided: Vector[][] = []; // 分割后的多边形数组

        let longestSideLength = 0;
        let longestSide = [p[0], p[1]];

        let perimeter = 0;

        for (let i = 0; i < p.length; i++) {
            const sideLength = p[i].clone().sub(p[(i+1) % p.length]).length();
            perimeter += sideLength;
            if (sideLength > longestSideLength) {
                longestSideLength = sideLength;
                longestSide = [p[i], p[(i+1) % p.length]];
            }
        }

        // 形状指数
        if (area / (perimeter * perimeter) < 0.04) {
            return [];
        }

        if (area < 2 * minArea) {
            return [p];
        }

        // 偏移量在 0.4 到 0.6 之间
        const deviation = (Math.random() * 0.2) + 0.4;

        const averagePoint = longestSide[0].clone().add(longestSide[1]).multiplyScalar(deviation);
        const differenceVector = longestSide[0].clone().sub(longestSide[1]);
        const perpVector = (new Vector(differenceVector.y, -1 * differenceVector.x))
            .normalize()
            .multiplyScalar(100);

        const bisect = [averagePoint.clone().add(perpVector), averagePoint.clone().sub(perpVector)];

        // 分割多边形
        try {
            const sliced = PolyK.Slice(PolygonUtil.polygonToPolygonArray(p), bisect[0].x, bisect[0].y, bisect[1].x, bisect[1].y);
            for (const s of sliced) {
                divided.push(...PolygonUtil.subdividePolygon(PolygonUtil.polygonArrayToPolygon(s), minArea));
            }

            return divided;
        } catch (error) {
            log.error(error);
            return [];
        }
    }

    /**
     * 缩放多边形
     * @param geometry 多边形顶点数组
     * @param spacing 缩放距离
     * @param isPolygon 是否为多边形
     */
    public static resizeGeometry(geometry: Vector[], spacing: number, isPolygon=true): Vector[] {
        try {
            const jstsGeometry = isPolygon? PolygonUtil.polygonToJts(geometry) : PolygonUtil.lineToJts(geometry);
            const resized = jstsGeometry.buffer(spacing, undefined, (jsts as any).operation.buffer.BufferParameters.CAP_FLAT);
            if (!resized.isSimple()) {
                return [];
            }
            return resized.getCoordinates().map(c => new Vector(c.x, c.y));
        } catch (error) {
            log.error(error);
            return [];
        }
    }

    /**
     * 计算多边形的平均点
     * @param polygon 多边形顶点数组
     */
    public static averagePoint(polygon: Vector[]): Vector {
        if (polygon.length === 0) return Vector.zeroVector();
        const sum = Vector.zeroVector();
        for (const v of polygon) {
            sum.add(v);
        }
        return sum.divideScalar(polygon.length);
    }

    /**
     * 判断点是否在多边形内
     * @param point 点
     * @param polygon 多边形顶点数组
     */
    public static insidePolygon(point: Vector, polygon: Vector[]): boolean {
        // ray-casting algorithm based on
        // http://www.ecse.rpi.edu/Homepages/wrf/Research/Short_Notes/pnpoly.html

        if (polygon.length === 0) {
            return false;
        }

        let inside = false;
        for (let i = 0, j = polygon.length - 1; i < polygon.length; j = i++) {
            const xi = polygon[i].x, yi = polygon[i].y;
            const xj = polygon[j].x, yj = polygon[j].y;

            const intersect = ((yi > point.y) != (yj > point.y))
                && (point.x < (xj - xi) * (point.y - yi) / (yj - yi) + xi);
            if (intersect) inside = !inside;
        }

        return inside;
    }

    /**
     * 判断点是否在矩形内
     * @param point 点
     * @param origin 矩形的起点
     * @param dimensions 矩形的宽高
     */
    public static pointInRectangle(point: Vector, origin: Vector, dimensions: Vector): boolean {
        return point.x >= origin.x && point.y >= origin.y && point.x <= dimensions.x && point.y <= dimensions.y;
    }

    /**
     * 将线转换为 JSTS 几何对象
     * @param line 线的顶点数组
     */
    private static lineToJts(line: Vector[]): jsts.geom.LineString {
        const coords = line.map(v => new jsts.geom.Coordinate(v.x, v.y));
        return PolygonUtil.geometryFactory.createLineString(coords);
    }

    /**
     * 将多边形转换为 JSTS 几何对象
     * @param polygon 多边形顶点数组
     */
    private static polygonToJts(polygon: Vector[]): jsts.geom.Polygon {
        const geoInput = polygon.map(v => new jsts.geom.Coordinate(v.x, v.y));
        geoInput.push(geoInput[0]);  // Create loop
        return PolygonUtil.geometryFactory.createPolygon(PolygonUtil.geometryFactory.createLinearRing(geoInput), []);
    }

    /**
     * 将多边形转换为数组
     * @param p 多边形顶点数组
     */
    private static polygonToPolygonArray(p: Vector[]): number[] {
        const outP: number[] = [];
        for (const v of p) {
            outP.push(v.x);
            outP.push(v.y);
        }
        return outP;
    }

    /**
     * 将数组转换为多边形
     * @param p 多边形顶点数组
     */
    private static polygonArrayToPolygon(p: number[]): Vector[] {
        const outP = [];
        for (let i = 0; i < p.length / 2; i++) {
            outP.push(new Vector(p[2*i], p[2*i + 1]));
        }
        return outP;
    }
}

File: src/py/polygon_util.py
import math
import random
from shapely.geometry import Polygon, LineString, Point
from shapely.ops import polygonize
from vector import Vector

class PolygonUtil:
    """多边形处理工具类"""
    
    @staticmethod
    def slice_rectangle(origin, world_dimensions, p1, p2):
        """通过线切割矩形，返回最小多边形"""
        rectangle = [
            [origin.x, origin.y],
            [origin.x + world_dimensions.x, origin.y],
            [origin.x + world_dimensions.x, origin.y + world_dimensions.y],
            [origin.x, origin.y + world_dimensions.y]
        ]
        
        rect_poly = Polygon(rectangle)
        line = LineString([(p1.x, p1.y), (p2.x, p2.y)])
        
        # 使用切割线切开矩形
        split_polys = []
        
        # 使用shapely的polygonize函数
        sliced_polys = list(polygonize(rect_poly.boundary.union(line)))
        
        if len(sliced_polys) < 2:
            # 如果切割失败，返回原始矩形
            return [Vector(x, y) for x, y in rectangle]
            
        # 找到面积最小的多边形
        min_area = float('inf')
        min_poly = None
        
        for poly in sliced_polys:
            area = poly.area
            if area < min_area:
                min_area = area
                min_poly = poly
                
        return [Vector(x, y) for x, y in list(min_poly.exterior.coords)[:-1]]
    
    @staticmethod
    def line_rectangle_polygon_intersection(origin, world_dimensions, line):
        """创建海域多边形"""
        if len(line) < 2:
            return []
            
        bounds = [
            (origin.x, origin.y),
            (origin.x + world_dimensions.x, origin.y),
            (origin.x + world_dimensions.x, origin.y + world_dimensions.y),
            (origin.x, origin.y + world_dimensions.y)
        ]
        
        bounding_poly = Polygon(bounds)
        line_shape = LineString([(p.x, p.y) for p in line])
        
        # 联合边界和线
        union = bounding_poly.exterior.union(line_shape)
        
        # 多边形化
        polygons = list(polygonize(union))
        
        if not polygons:
            return []
            
        # 找到最小面积的多边形
        smallest_poly = min(polygons, key=lambda p: p.area)
        
        return [Vector(x, y) for x, y in list(smallest_poly.exterior.coords)[:-1]]
    
    @staticmethod
    def calc_polygon_area(polygon):
        """计算多边形面积"""
        if len(polygon) < 3:
            return 0
            
        total = 0
        
        for i in range(len(polygon)):
            add_x = polygon[i].x
            add_y = polygon[(i + 1) % len(polygon)].y
            sub_x = polygon[(i + 1) % len(polygon)].x
            sub_y = polygon[i].y
            
            total += (add_x * add_y * 0.5)
            total -= (sub_x * sub_y * 0.5)
            
        return abs(total)
    
    @staticmethod
    def subdivide_polygon(p, min_area):
        """递归地沿着最长边分割多边形，直到满足最小面积停止条件"""
        area = PolygonUtil.calc_polygon_area(p)
        
        if area < 0.5 * min_area:
            return []
            
        divided = []
        
        # 找最长边
        longest_side_length = 0
        longest_side = [p[0], p[1]]
        
        perimeter = 0
        
        for i in range(len(p)):
            side_length = p[i].clone().sub(p[(i+1) % len(p)]).length()
            perimeter += side_length
            
            if side_length > longest_side_length:
                longest_side_length = side_length
                longest_side = [p[i], p[(i+1) % len(p)]]
        
        # 形状指数，使用1:4矩形比例作为限制
        if area / (perimeter * perimeter) < 0.04:
            return []
            
        if area < 2 * min_area:
            return [p]
            
        # 介于0.4和0.6之间
        deviation = (random.random() * 0.2) + 0.4
        
        average_point = longest_side[0].clone().add(longest_side[1]).multiply_scalar(deviation)
        difference_vector = longest_side[0].clone().sub(longest_side[1])
        perp_vector = Vector(difference_vector.y, -1 * difference_vector.x)
        perp_vector.normalize().multiply_scalar(100)
        
        bisect = [average_point.clone().add(perp_vector), average_point.clone().sub(perp_vector)]
        
        # 使用Shapely执行分割
        try:
            poly = Polygon([(v.x, v.y) for v in p])
            cut_line = LineString([(bisect[0].x, bisect[0].y), (bisect[1].x, bisect[1].y)])
            
            # 将多边形分割成两部分
            sliced = []
            for geom in PolygonUtil._split(poly, cut_line):
                if isinstance(geom, Polygon):
                    sliced.append([Vector(x, y) for x, y in list(geom.exterior.coords)[:-1]])
            
            # 递归调用
            for s in sliced:
                divided.extend(PolygonUtil.subdivide_polygon(s, min_area))
                
            return divided
        except Exception as e:
            print(f"多边形分割错误: {e}")
            return []
    
    @staticmethod
    def _split(polygon, line):
        """辅助函数，用于分割多边形"""
        merged = polygon.boundary.union(line)
        borders = polygonize(merged)
        return list(borders)
    
    @staticmethod
    def resize_geometry(geometry, spacing, is_polygon=True):
        """收缩或扩展多边形"""
        try:
            if is_polygon:
                if len(geometry) < 3:
                    return []
                geom = Polygon([(v.x, v.y) for v in geometry])
            else:
                if len(geometry) < 2:
                    return []
                geom = LineString([(v.x, v.y) for v in geometry])
                
            # 使用buffer进行扩展或收缩
            resized = geom.buffer(spacing, cap_style=1)
            
            if not resized.is_simple:
                return []
                
            if is_polygon:
                return [Vector(x, y) for x, y in list(resized.exterior.coords)[:-1]]
            else:
                return [Vector(x, y) for x, y in list(resized.coords)]
        except Exception as e:
            print(f"几何调整错误: {e}")
            return []
    
    @staticmethod
    def average_point(polygon):
        """计算多边形的平均点"""
        if not polygon:
            return Vector.zero_vector()
            
        sum_point = Vector.zero_vector()
        for v in polygon:
            sum_point.add(v)
            
        return sum_point.divide_scalar(len(polygon))
    
    @staticmethod
    def inside_polygon(point, polygon):
        """判断点是否在多边形内部"""
        if not polygon:
            return False
        
        # 射线投射算法
        inside = False
        for i in range(len(polygon)):
            j = (i - 1) % len(polygon)
            
            xi, yi = polygon[i].x, polygon[i].y
            xj, yj = polygon[j].x, polygon[j].y
            
            intersect = ((yi > point.y) != (yj > point.y)) and \
                       (point.x < (xj - xi) * (point.y - yi) / (yj - yi) + xi)
                       
            if intersect:
                inside = not inside
                
        return inside
    
    @staticmethod
    def point_in_rectangle(point, origin, dimensions):
        """判断点是否在矩形内部"""
        return (point.x >= origin.x and point.y >= origin.y and 
                point.x <= dimensions.x and point.y <= dimensions.y)